Create Windows installer

kthchew opened the pull request:
Uses NSIS to create a Windows installer. (I chose NSIS because there are also tools on Linux/macOS to make NSIS installers, so it's easier to build locally.) Closes #66. 

Things to test:

- Installing with the GUI, both the non-portable and portable versions (they use the same installer)
- Installing in silent mode (`\path\to\PolyMC-Setup.exe /S`) (other command line options at https://nsis.sourceforge.io/Docs/Chapter3.html under section 3.2)
- Uninstalling
- Installing/uninstalling to non-default locations

Explodey-Wolf commented:
Yes. This would be great.

ZekeZDev commented:
Looks great!

Scrumplex commented on .github/workflows/build.yml:
```diff
@@ -190,6 +191,13 @@ jobs:
+        if: runner.os == 'Windows'
+        shell: msys2 {0}
+        run: |
+          cd ${{ env.INSTALL_PORTABLE_DIR }}
```
Why are you packaging the portable version into an installer. This would cause the data folder to be where PolyMC is installed. This should probably rather be the non-portable version

kthchew replied:
At install time, there’s an optional checkbox that’s labeled “portable” which is deselected by default. The portable component is only installed if the user selects that checkbox; by default, it installs the non-portable version. I have to use the portable directory in order to get the `portable.txt`.

Scrumplex replied:
Oh hmm interesting. What will be the target location if the user selects the portable version?

kthchew replied:
It’s the same location by default. Although if they select the portable version, I’d assume they’d change the install location when asked to do so.

Scrumplex commented on .github/workflows/build.yml:
```diff
@@ -257,6 +265,13 @@ jobs:
+        if: runner.os == 'Windows'
+        uses: actions/upload-artifact@v3
+        with:
+          name: PolyMC-${{ matrix.name }}-${{ env.VERSION }}-${{ inputs.build_type }}-Setup
```
Let's keep the same naming scheme

```suggestion
          name: PolyMC-${{ matrix.name }}-Setup-${{ env.VERSION }}-${{ inputs.build_type }}
```

Scrumplex commented on program_info/win_install.nsi:
```diff
@@ -0,0 +1,223 @@
+
+SectionEnd
+
+Section /o "Portable"
+
+  SetOutPath $INSTDIR
+  File "portable.txt"
+
+SectionEnd
```
I feel like the `Portable` checkbox might cause confusion, as its consequences aren't explained anywhere. It also doesn't make much sense in this case, as it probably doesn't make sense registering a portable application like any installed application.

kthchew replied:
I did that originally so that portable installs would be able to use that installer to update themselves, since most Windows update libraries need an installer to run. But then I realized that even installing the “regular” version on top won’t affect the `portable.txt`, so that shouldn’t be a problem.

What might be a problem is that there may need to be a workaround for when portable PolyMC installations want to do an update by running the installer. It would register them as an installed app upon an update in the installer’s current state, even if they weren’t registered as such before.

Edit: although this portion of code doesn't do anything about it. I may just remove the portable checkbox, in that case.

Scrumplex commented:
This should probably also be added to the release workflow.

kthchew commented:
I removed the portable checkbox, and added two new command line options:

- `/NoUninstaller`, which does not write any of the uninstaller registry keys or the uninstaller itself, intended for use while updating a portable install
- `/NoShortcuts`, which disables the start menu shortcuts component, intended for use while updating any (regular and portable) installs

Scrumplex commented on .github/workflows/trigger_release.yml:
```diff
@@ -43,10 +43,12 @@ jobs:
             test -z "${PORT}" || NAME="${NAME}-Portable"
-            zip -r -9 "../${NAME}-${{ env.VERSION }}.zip" *
+            test -z "${INST}" || mv PolyMC-*.exe ../${NAME}-Setup-${{ env.VERSION }}.exe
+            test -n "${INST}" || zip -r -9 "../${NAME}-${{ env.VERSION }}.zip" *
```
I think we should also provide the zip file in the GH release, even if we don't necessarily link to it on the website.

kthchew replied:
For clarification, do you mean the zip files that have always been made (the regular and portable ones)? They should still be made and attached to the release. If they are not, that was a mistake.

Scrumplex replied:
Oh, I see. Just a brain fart again. lol

kthchew replied:
No worries, shell syntax with these `test` commands is a little weird 😅

Scrumplex reviewed (approved):
Looks fine overall. Thanks!

Un1q32 commented:
@kthchew using / in command line options isnt generally used anymore.  you only see it in old software.  we should be using  - like everyone else

kthchew commented:
> @kthchew using / in command line options isnt generally used anymore. you only see it in old software. we should be using - like everyone else

The silent parameter which is added by default by NSIS only accepts `/S`, not `-S`. It would be weird to specify the silent parameter with a `/` but everything else with a `-` in the same command.

DioEgizio commented:
seems to work fine from my testing!

Un1q32 commented:
> > @kthchew using / in command line options isnt generally used anymore. you only see it in old software. we should be using - like everyone else
> 
> The silent parameter which is added by default by NSIS only accepts `/S`, not `-S`. It would be weird to specify the silent parameter with a `/` but everything else with a `-` in the same command.

so its an issue with NSIS then.  got it

rollingmoai commented:
It would be great if it could be installed system-wide (https://github.com/PolyMC/PolyMC/issues/978).
